src/services/user-config-service.ts

```diff
@@ -0,0 +1,120 @@
+
+const USER_CONFIG_FILE = ".jira/config.yml";
+const logger = getLogger("services.user-config");
+const MAX_PATTERNS_PER_ENVIRONMENT = 10;
+
+/**
+ * Checks whether a list of modified files contains the config file. If yes, reads that config file
+ * from the GitHub repository, parses it, and stores the config against the given repository
+ * in the database.
+ *
+ * This function is meant to be called whenever there is a change in the repository so we can check
+ * if the config file has changed.
+ *
+ * @param subscription the subscription to which the repository belongs.
+ * @param repositoryId the ID of the repository.
+ * @param githubInstallationId the ID of the installation to which the repository belongs.
+ * @param modifiedFiles list of modified files (added, modified, or removed). The config will only be updated if this list contains
+ * the config file.
+ */
+export const updateRepoConfig = async (
+	subscription: Subscription,
+	repositoryId: number,
+	githubInstallationId: InstallationId,
+	modifiedFiles: string[] = []): Promise<void> => {
+
+	if (modifiedFiles.includes(USER_CONFIG_FILE)) {
+		try {
+			const repoSyncState = await RepoSyncState.findByRepoId(subscription, repositoryId);
+			await updateRepoConfigFromGitHub(repoSyncState, githubInstallationId);
+		} catch (err) {
+			logger.error({
+				err,
+				githubInstallationId,
+				repositoryId
+			}, "error while updating the repo config");
+		}
+	}
+};
+
+/**
+ * Returns the config for a given repo.
+ */
+export const getRepoConfig = async (subscription: Subscription, repositoryId: number): Promise<Config | undefined> => {
+	// In the future, we may look in other places for a config than just in the RepoSyncState (for example,
+	// we might fall back to default configs on the level of a subscription or an installation).
+	const repoSyncState = await RepoSyncState.findByRepoId(subscription, repositoryId);
+	return repoSyncState.config;
+};
+
+/**
+ * Fetches contents from CONFIG_PATH from GitHub via GitHub's API, transforms it from base64 to ascii and returns the transformed string.
+ */
+const getRepoConfigFromGitHub = async (githubInstallationId: InstallationId, owner: string, repo: string): Promise<string | undefined> => {
+	const client = new GitHubInstallationClient(githubInstallationId, logger);
+	const contents = await client.getRepositoryFile(owner, repo, USER_CONFIG_FILE);
+
+	if (!contents) {
+		return undefined;
+	}
+	return Buffer.from(contents, "base64").toString("utf-8");
+};
+
+/**
+ * Iterates through environment patterns and returns true if any environment contains too many patterns to test against.
+ */
+const hasTooManyPatternsPerEnvironment = (config: Config): boolean => {
+	const environmentMapping = config?.deployments?.environmentMapping;
+	if (!environmentMapping) {
+		return false;
+	}
+	return Object.keys(environmentMapping).some(key => {
+		return environmentMapping[key]?.length > MAX_PATTERNS_PER_ENVIRONMENT;
+	});
```

Comment:
What's this for?

Reply:
just a check to limit user input to not allow any nonsense